File: Review-03-Separation-of-Concerns/Example-6-Java/src/main/java/UpdateRoom.java
import java.io.BufferedReader;
import java.io.StringReader;
import java.io.IOException;
import java.util.Scanner;

import house.Room;
import house.House;

public class UpdateRoom
{

    public static final String ROOM_DATA = "Laundry Room; 8 4 1.95 Laminate\n"
            + "Kitchen; 20 12 3.87 Tile\n"
            + "Storage Room; 16 16 4.39 Birch Wood";

    /**
     * Build our example house
     */
    static void buildHouse(BufferedReader reader, House house)
        throws IOException
    {
        String line;

        while ((line = reader.readLine()) != null) {
            int idxSemicolon = line.indexOf(';');

            String name = line.substring(0, idxSemicolon);
            String theRest = line.substring(idxSemicolon + 1, line.length());

            Scanner withinLineScanner = new Scanner(theRest);
            double length = withinLineScanner.nextDouble();
            double width = withinLineScanner.nextDouble();
            double unitCost = withinLineScanner.nextDouble();
            String type = withinLineScanner.nextLine();

            Room aRoom = new Room(name,
                                  new Room.DimensionSet(length, width),
                                  unitCost,
                                  type);

            house.addRoom(aRoom);
        }
    }

    /**
     * Take a room and change the flooring
     *
     * @param original House to change
     *
     * @return House with the updated flooring
     */
    static House upgradeFlooring(House original)
    {
        House modified = new House();
        modified.setName("After Stone Bricks");

        for (Room rm : original) {
            Room copy = rm.clone();
            copy.setFlooring("Stone Bricks", 12.97);

            modified.addRoom(copy);
        }

        return modified;
    }

    /**
     * Compute the discounting flooring price for a single Room.
     *
     * @param r room to examine
     * @param percent discount as a decimal
     *
     * @pre 0 <= percent && percent <= 1
     */
    static double discountFlooring(Room r, double percent)
    {
        final double scale = 1 - percent;

        return scale * r.flooringCost();
    }

    /**
     * Compute the area of a room and the cost of
     * flooring for the room
     * <p>
     * Let us Review the use of reference and pointer variables.
     * <p>
     * We will use these when we implement the iterator interface.
     */
    public static void main(String... args)
        throws IOException
    {
        // Construct, build, and print a house
        House house = new House();

        BufferedReader fakeInputFile = new BufferedReader(new StringReader(ROOM_DATA));
        buildHouse(fakeInputFile, house);

        System.out.println(house);
        System.out.println();

        // Upgrade the flooring in a second duplicate house
        House duplicateHouse = upgradeFlooring(house);

        System.out.printf(
            "house == duplicateHouse     -> %b%n",
             (house.equals(duplicateHouse))
        );
        System.out.printf(
            "&house == &duplicateHouse   -> %b%n",
             (house == duplicateHouse)
        );
        System.out.println();

        System.out.println(house);
        System.out.println();
        System.out.println(duplicateHouse);
        /*

        // Get all the flooring costs with a 10% discount
        auto discountFunc = std::bind(discountFlooring, std::placeholders::_1, 0.1);

        vector<double> costs(duplicateHouse.size());
        std::transform(duplicateHouse.begin(), duplicateHouse.end(), costs.begin(),
                       discountFunc);

        std::copy(costs.begin(), costs.end(),
                  std::ostream_iterator<double>(std::cout, "\n"));

        // Print the sum, min, max -> D.R.Y!
        cout << "Total: "
             << std::accumulate(costs.begin(), costs.end(), 0.0, std::plus<double>())
             << "\n";
        cout << "Min: "
             << *std::min_element(costs.begin(), costs.end())
             << "\n"
             << "Max: "
             << *std::max_element(costs.begin(), costs.end())
             << "\n";
        */

        // I would probably use minmax_element and auto
        /*
        std::pair<std::vector<double>::const_iterator,
                  std::vector<double>::const_iterator> extremes = std::minmax_element(costs.begin(), costs.end());
        */

        /*
        auto extremes = std::minmax_element(costs.begin(), costs.end());

        cout << "Min: " << *(extremes.first)  << "\n"
             << "Max: " << *(extremes.second) << "\n";

        return 0;
        */
    }
}

File: Review-14-Python-Builder-Pattern/Example-1/update_room.py
#! /usr/bin/env python

import copy

from house import House
from room import Room, RoomBuilder


def main():
    """
    Compute the area of a room and the cost of
    flooring for the room

    Let us Review the use of reference and pointer variables.

    We will use these when we implement the iterator interface.
    """

    # Construct, build, and print a house
    house = House()
    build_house(house)
    print(house)

    # Upgrade the flooring in a second duplicate house
    duplicate_house = upgrade_flooring(house)

    print(f"house == duplicate_house -> {house == duplicate_house}")
    print(f"&house == &duplicate_house -> {house is duplicate_house}")
    print()

    print(house)
    print(duplicate_house)

    # Get all the flooring costs with a 10% discount
    costs = [discount_flooring(room) for room in duplicate_house]
    #  costs = []
    #  for room in duplicate_house:
    #      costs.append(discount_flooring(room))

    for room_cost in costs:
        print(f"{room_cost:.2f}")

    # Print the sum, min, max -> D.R.Y!
    total = sum(costs)
    min_c = min(costs)
    max_c = max(costs)

    print(f"Total: {total:.2f}")
    print(f"Min  : {min_c:.2f}")
    print(f"Max  : {max_c:.2f}")


def build_house(house: House) -> None:
    """
    Build our example house
    """

    house.add_room(
        RoomBuilder()
        .with_name("Laundry Room")
        .with_dimensions(8, 4)
        .with_flooring("Laminate", 1.95)
        .build()
    )

    kitchen = (
        RoomBuilder()
        .with_name("Kitchen")
        .with_dimensions(20, 12)
        .with_flooring("Tile", 3.87)
        .build()
    )

    house.add_room(kitchen)

    house.add_room(
        RoomBuilder()
        .with_name("Storage Room")
        .with_dimensions(16, 16)
        .with_flooring("Birch Wood", 4.39)
        .build()
    )


def upgrade_flooring(original: House) -> House:
    """
    Take a room and change the flooring

    Args:
        original: House to change

    Returns:
        House with the updated flooring
    """

    modified = copy.deepcopy(original)

    for room in modified:
        room.set_flooring("Stone Bricks", 12.97)

    #  modified.set_name("After Stone Bricks")
    modified.name = "After Stone Bricks"  # Magic!!

    return modified


def discount_flooring(a_room: Room) -> float:
    """
    Take a room, discount the flooring cost by 90%.

    Returns:
        Discounted flooring cost
    """

    return 0.90 * a_room.flooring_cost()


if __name__ == "__main__":
    main()
